tests/src/kernel/hipLaunchParm.cpp

```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
+  char c1;
+  short int si;
+} __attribute__((aligned(8)))  hipLaunchKernelStruct_t6;
+
+// This test is to verify struct with aligned(16),
+// right now it's brokenon hcc & hip-clang
+typedef struct hipLaunchKernelStruct7 {
+  char c1;
+  short int si;
+} __attribute__((aligned(16)))  hipLaunchKernelStruct_t7;
+
+// This test is to verify struct with packed & aligned,
+// size should be 4Bytes right now it's broken on hcc & hip-clang
+typedef struct hipLaunchKernelStruct8 {
+  char c1;
+  short int si;
+  bool b;
+}__attribute__((packed, aligned(4)))  hipLaunchKernelStruct_t8;
+
+// This test is to verify const struct object
+typedef struct hipLaunchKernelStruct9 {
+  char c1;
+  uint32_t* ip;  // uint pointer
+} hipLaunchKernelStruct_t9;
+
+// This test is to verify struct with stdint types, uintN_t
+typedef struct hipLaunchKernelStruct10 {
+  uint64_t u64;
+  uint32_t u32;
+  uint8_t u8;
+} hipLaunchKernelStruct_t10;
+
+// This test is to verify struct with volatile member
+typedef struct hipLaunchKernelStruct11 {
+  int i1;
+  volatile unsigned int vint;
+} hipLaunchKernelStruct_t11;
+
+// This test is to verify struct with simple class object
+class base {
+ public:
+    int i = 0;
+    base() {}
+};
+typedef struct hipLaunchKernelStruct12 {
+  base b;
+  char c1;
+} hipLaunchKernelStruct_t12;
+
+// This test is to verify struct with __device__ func() attribute
+typedef struct hipLaunchKernelStruct13 {
+  int i1;
+  __device__ int getvalue() { return i1; }
+} hipLaunchKernelStruct_t13;
+
+// This test is to verify struct with array variable,
+// write to from device
+typedef struct hipLaunchKernelStruct14 {
+  int readint;
+  int writeint[BLOCK_DIM_SIZE];  // will write to this from device
+} hipLaunchKernelStruct_t14;
+
+// This test is to verify struct with dynamic memory, new int
+// the heap memory will be accessed from device
+typedef struct hipLaunchKernelStruct15 {
+  char c1;
+  int* heapmem = new int[BLOCK_DIM_SIZE];
+} hipLaunchKernelStruct_t15;
+
+// This test is to verify simple template struct
+template<typename T>
+struct hipLaunchKernelStruct_t16 {
+  T t1;
+};
+
+// This test is to verify simple explicity template struct
+template<typename T> struct hipLaunchKernelStruct_t17 {};
+template<>  // explicit template
+struct hipLaunchKernelStruct_t17<int> {
+  int t1;
+};
+
+// This test is to verity write to struct memory using __device__ func()
+typedef struct hipLaunchKernelStruct18 {
+  char c1;
+  __device__ void setChar(char c) { c1 = c; }
+  __device__ int getChar() { return c1; }
+} hipLaunchKernelStruct_t18;
+
+// This test is to verity user defined STL, simple stack implementation
+typedef struct stackNode {
+    int data;
+    stackNode* nextNode = NULL;
+} stackNode_t;
+typedef struct hipLaunchKernelStruct19 {
+  stackNode_t* stack = NULL;
+  unsigned int size_ = 0;
+  void pushMe(int value) {  // not a device function, setting from host
+    stackNode_t* newNode = new stackNode_t;
```

Comment:
device code cannot access host heap memory, so this does not work.
For HCC, there is device function `__hip_hc_malloc` and `__hip_hc_free`, which you may use to implement device function pushMe and popMe. For host code, you may want to use hipMalloc and hipFree.

Reply:
Done